zebra-chain/src/transaction/serialize.rs

```diff
@@ -200,13 +200,42 @@ impl ZcashSerialize for Transaction {
                 // Write version 5 and set the fOverwintered bit.
                 writer.write_u32::<LittleEndian>(5 | (1 << 31))?;
                 writer.write_u32::<LittleEndian>(TX_V5_VERSION_GROUP_ID)?;
+
                 lock_time.zcash_serialize(&mut writer)?;
                 writer.write_u32::<LittleEndian>(expiry_height.0)?;
                 inputs.zcash_serialize(&mut writer)?;
                 outputs.zcash_serialize(&mut writer)?;
 
-                // TODO: serialize sapling shielded data according to the V5 transaction spec
+                match sapling_shielded_data {
```

Comment:
Please follow the field order in the v5 transaction spec.

Spends have 1 count (compactsize), but they are split into 3 separate arrays in 2 different locations.
Outputs have 1 count, but they are split into 2 separate arrays in 2 different locations.

<img width="827" alt="Screen Shot 2021-04-15 at 09 51 18" src="https://user-images.githubusercontent.com/8951843/114796893-55528b80-9dd5-11eb-9763-9d0713bd37d9.png">

Reply:
Also, we should move the serialize and deserialize implementations to `Option<sapling::ShieldedData<SharedAnchor>>`, to match the design:
https://github.com/ZcashFoundation/zebra/pull/1996/files#diff-a377e69ec608cdeb4b2c5bf8159152daf73ad6584e960882525828bf79bfdc96R277

So v5 sapling serialization will look like the Sprout JoinSplit serialization, where the code is in a separate function from v5 transaction serialization:
https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/transaction/serialize.rs#L34
https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/transaction/serialize.rs#L124
https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/transaction/serialize.rs#L247

Let's also serialize and deserialize `Option<sapling::ShieldedData<SharedAnchor>>` for consistency.